File: 1-500/0008. String to Integer (atoi).java
// Medium
// Math
// Time:O(n),Space:O(1)
// https://leetcode.cn/problems/string-to-integer-atoi/

class Solution {
  // Skip leading whitespace, and handle optional sign('+'or'-')
  // Parse digits and accumulate res, then check overflow 
  public int myAtoi(String s) {
      int index = 0, sign = 1, num = 0, n = s.length();
      while (index < n && s.charAt(index) == ' ') index++;

      if (index < n && (s.charAt(index) == '+' || s.charAt(index) == '-')) {
          sign = (s.charAt(index) == '-') ? -1 : 1;
          index++;
      }

      while (index < n && Character.isDigit(s.charAt(index))) {
          int digit = s.charAt(index) - '0';
          // check overflow
          if (num > (Integer.MAX_VALUE - digit) / 10) {
              return (sign == 1) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
          }
          num = num * 10 + digit;
          index++;
      }
      return num * sign;
  }
}

File: 0001-0500/0008. String to Integer (atoi).py
# Time:O(n), Space:O(1)

class Solution:
    def myAtoi(self, s: str) -> int:
        i = 0
        n = len(s)

        while i < n and s[i] == " ":
            i += 1

        if i == n:
            return 0

        sign = 1
        if s[i] == "-":
            sign = -1
            i += 1
        elif s[i] == "+":
            i += 1

        res = 0
        while i < n and s[i].isdigit():
            digit = int(s[i])

            if res > (2**31 - 1) // 10 or (res == (2**31 - 1) // 10 and digit > 7):
                return 2**31 - 1 if sign == 1 else -(2**31)

            res = res * 10 + digit
            i += 1

        return sign * res
